Build failure: linuxKernel.kernels.linux_5_10

### Nixpkgs version

- Unstable (25.05)

### Steps to reproduce

`nix build -f channel:nixos-unstable linuxKernel.kernels.linux_5_10`

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/296082778

### Relevant log output

```console
GEN     .version
  CHK     include/generated/compile.h
  LD      vmlinux.o
  MODPOST vmlinux.symvers
  MODINFO modules.builtin.modinfo
  GEN     modules.builtin
  LD      .tmp_vmlinux.btf
  BTF     .btf.vmlinux.bin.o
  LD      .tmp_vmlinux.kallsyms1
  KSYMS   .tmp_vmlinux.kallsyms1.S
  AS      .tmp_vmlinux.kallsyms1.o
  LD      .tmp_vmlinux.kallsyms2
  KSYMS   .tmp_vmlinux.kallsyms2.S
  AS      .tmp_vmlinux.kallsyms2.o
  LD      vmlinux
  BTFIDS  vmlinux
FAILED unresolved symbol filp_close
make[1]: *** [/build/linux-5.10.237/Makefile:1215: vmlinux] Error 255
make: *** [../Makefile:192: __sub-make] Error 2
```

### Additional context

It's broken since the [update to v5.10.236](https://github.com/NixOS/nixpkgs/commit/109b3a01dbeeb78483b52937420385ecf8f2cf1f).  This was already [noticed in the updating pull request](https://github.com/NixOS/nixpkgs/pull/397717#issuecomment-2798850343), and a [fix was expected](https://github.com/NixOS/nixpkgs/pull/397717#issuecomment-2798852628).  However, [the update to v5.10.237](https://github.com/NixOS/nixpkgs/pull/403572) didn't help, and [it's also broken on current `staging-next`](https://hydra.nixos.org/build/296090083).

I bisected the kernel sources to find the root cause:  Apparently, reverting these *two* commits

* https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=51c36f0d211ec4d5a9747947c5575336004bf3d8 (part of v5.10.236)
* https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=9dbf002e7dacd6536cb47c388512dbd3b33b6313 (part of v5.10.237)

fixes v5.10.237 .  Reverting the first one fixes v5.10.236.  However, from looking at the error message and into the commit diffs, I can absolutely not grasp what's going on here.  In particular, searching the internet for `filp_close` and one of those commit titles returns nothing.

These are two interesting correspondences on the internet I found while researching the error message, with some observations/statements by Greg Kroah-Hartman -- unfortunatelly, the problem seems to be considered solved:

* https://www.spinics.net/lists/stable/msg821695.html
* https://lore.kernel.org/all/b39ca723-16a4-42a2-b8ca-b97d0e4bf7f5@manjaro.org/T/

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.88, NixOS, 24.11 (Vicuna), 24.11.717586.3e362ce63e16`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.14`
 - channels(root): `"nixos-24.11"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

### Notify maintainers

Notifying linux kernel maintainer team @TredwellGit , @K900 , @Ma27 , @NeQuissimus , @alyssais

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/os-specific/linux/kernel/common-config.nix
+++ b/pkgs/os-specific/linux/kernel/common-config.nix
@@ -1102,7 +1102,10 @@ let
 
         # enable support for device trees and overlays
         OF = option yes;
-        OF_OVERLAY = option yes;
+        # OF_OVERLAY breaks v5.10 on x86_64, see https://github.com/NixOS/nixpkgs/issues/403985
+        OF_OVERLAY = lib.mkIf (!(lib.versionOlder version "5.15" && stdenv.hostPlatform.isx86_64)) (
+          option yes
+        );
 
         # Enable initrd support.
         BLK_DEV_INITRD = yes;

